### PR TITLE
Add Tridiagonal construction rule

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/SciML/DiffEqFlux.jl/issues/744

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
@avik-pal

**Review by avik-pal — APPROVED** [review]

File: test/rulesets/LinearAlgebra/structured.jl [hunk 1]
```diff
@@ -161,4 +161,9 @@
             end
         end
     end
+
+    @testset "Tridiagonal" begin
+        res, pb = rrule(Tridiagonal, [1, 4], [2, 3, 4], [5, 3])
```
Comment (oxinabox) [inline test/rulesets/LinearAlgebra/structured.jl:166]: Can this be written usually ChainRulesTestUtils's test rule
I think it should be just 
```suggestion
        test_rrule(Tridiagonal, [1.0, 4.0], [2.0, 3.0, 4.0], [5.0, 3.0])
```

File: src/rulesets/LinearAlgebra/structured.jl [hunk 2]
```diff
@@ -268,3 +268,20 @@ function rrule(::typeof(logdet), X::Union{Diagonal, AbstractTriangular})
...
+#####
+
+function rrule(::Type{Tridiagonal}, dl, d, du)
+    y = Tridiagonal(dl, d, du)
+    @views function ∇Tridiagonal(∂y)
+        return (
+            NoTangent(),
+            diag(∂y[2:end, 1:(end - 1)]),
+            diag(∂y),
+            diag(∂y[1:(end - 1), 2:end]),
```
Comment (oxinabox) [inline src/rulesets/LinearAlgebra/structured.jl:283]: Can't this be
```suggestion
            diag(∂y, 1),
            diag(∂y),
            diag(∂y, -1),
```
Or maybe other order?

Comment (ChrisRackauckas) [reply src/rulesets/LinearAlgebra/structured.jl:283]: other order

Comment (ChrisRackauckas) [inline src/rulesets/LinearAlgebra/structured.jl:283]:
```suggestion
            diag(∂y, -1),
            diag(∂y),
            diag(∂y, 1),
```

**Review by oxinabox — APPROVED** [review]

**Comment by mcabbott** [comment]
Is there a minimal example of what problem this solves?  

There is some chance that it would be fixed by https://github.com/JuliaDiff/ChainRulesCore.jl/pull/446 , which is really part of the original idea of ProjectTo.

**Comment by ChrisRackauckas** [comment]
Not fixed just hit it here https://github.com/SciML/DiffEqFlux.jl/actions/runs/9115130596/job/25060691159?pr=922

**Comment by oxinabox** [comment]
Once https://github.com/JuliaDiff/FiniteDifferences.jl/pull/237
is merged this will pass tests.

**Comment by avik-pal** [comment]
What is the current status for this?

**Comment by oxinabox** [comment]
Merged.
It was just blocked waiting for me to come back and see it passed

**Pull request merged**